### ISSUE TITLE
[destinE] Add Dynamic user mapping

### ISSUE THREAD

**Issue by kinow** [BSC-ES/autosubmit#1454]
In GitLab by @dbeltrankyl on Nov 19, 2024, 09:35

Hello @kinow 

Just to keep track of it inside Autosubmit and set the correct milestone, I'm opening this issue to address https://earth.bsc.es/gitlab/digital-twins/de_340-2/workflow/-/issues/422

Thanks to @Nortamo:
 
The idea is to allow dynamic user mapping so we can have single user to launch testing experiments through Jenkins.

He discarded a solution using only paramiko due: https://earth.bsc.es/gitlab/digital-twins/de_340-2/workflow/-/issues/422#note_321356

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Nov 19, 2024, 09:36

@kinow Since you're busy with the review, maybe I can have a look ( although I need to look at the issue for more context ) after looking at my pending merges so the 4.1.12 release can come sooner. As you prefer

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Nov 19, 2024, 09:37

mentioned in issue digital-twins/de_340-2/workflow#422

**Comment by kinow** [comment]
Hi @dbeltrankyl ! It'd be really great if you could have a look as you have more experience with the networking part of platforms & Paramiko in general. If we solve this then Aina/Francesc/Leo/others will be really happy to have the testing suite automated.

From Henrik's message looks Paramiko didn't work with SSH config Match statements. Food for thought, but maybe one of these things could help:

- Paramiko is Python... so we can always monkey-patch it; i.e. it's not nice, but if what Henrik said that's missing in Paramiko is simple, we can always "patch it" (an elegant way to say Hack-It™).
- Not sure if there is another way to implement `Match` statements that could work with Paramiko (it's still a good thing that they are the first ones with Match statements, as we can learn how to use it in AS)

Cheers

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Nov 20, 2024, 12:55

Hello @kinow 

I just started this issue; I looked at all the context because I only commented on some small bits whenever I was mentioned. 

The problem now is that the connection to each platform is done with tsuite1@host? 

Original comment: 

> So the dynamic user mapping will have to be done in autosubmit, easiest way would be that you allow for an environment variable to be set before starting autosubmit where the usermapping is done, or if you then provide a safe way to pass in platforms.yml so that it's dynamically set based on the triggering user.


* [x] Platforms.yml variable

I think this is not possible. I thought we could add a %platforms.platform.mn5.shared_user% and then rely upon the user-defined platforms.yml, but since everyone will be tsuite1, we can't make it versioned as it won't read the one that each user have

* [ ] Enviroment variable and custom_config. 

Maybe this would be the easiest way of doing it. 

1) Change the minimal.yml custom_config for the test experiments: ( this would be done only the first time these experiments ran )

[dbeltran@climatedt-wf conf]$ cat minimal.yml 

```yaml
...
DEFAULT:
...
  #hint: use %PROJDIR% to point to the project folder (where the project is cloned)
  # Custom configuration location.
  CUSTOM_CONFIG: "%PROJDIR%/conf/bootstrap/"
...
```
to: 

```yaml
...
DEFAULT:
...
  #hint: use %PROJDIR% to point to the project folder (where the project is cloned)
  # Custom configuration location.
  CUSTOM_CONFIG: 
    PRE: "%PROJDIR%/conf/bootstrap/"
    POST: "%ENV_PLATFORMS_PATH%"
...
```
2) AutosubmitConfigParser: def reload(...):

```python
#code to load minimal.yml
platforms_path = os.getenv("ENV_PLATFORMS_PATH") # add code to check if exists etc
as_conf.experiment_data["ENV_PLATFORMS_PATH"] = platforms_path
# rest of the code
```

3) Set ENV_PLATFORMS_PATH

```bash
$ ssh $climate-dt
$ sudo -i -u $shareduser
$ module load autosubmit/4.1.12-dev... # <- future
$ export ENV_PLATFORMS_PATH="~/$your_real_user_platforms.yml"
$ autosubmit create -np -v a1um
$ autosubmit -lc DEBUG run a1um
```

4) Create the platforms.yml file ( this would be done only the first time these experiments ran )

```yaml
Platforms:
  marenostrum5:
    user: $your_real_user
```

what do you think? 
Edit:
Is there something missing? 

I understood that allowing, for example, SSH bsc032XXX@mn5 is handled outside AS , no?

**Comment by kinow** [comment]
>The problem now is that the connection to each platform is done with tsuite1@host?

Correct. Users sudo'ed to the testing suite user can connect to `lumi`, and the `~/.ssh/config` rules will match the request, and execute a shell script that contains a table of users and some logic to validate and connect with the correct user.

For example, I just `sudo`'ed as the testing suite user, and now the `$SUDO_USER` shows my actual user (which issued `sudo ...`).

The SSH `Match` is used on Linux to check if my user is allowed, and then find the remote user actual ID. The idea was that since that works well on the terminal, it would also work with Autosubmit, when an experiment tried to do the same to connect to a remote platform.

But apparently Paramiko is not happy with our use of Match (or not compliant to our version of Bash/SSH?).

>what do you think? Edit: Is there something missing?
>
>I understood that allowing, for example, SSH bsc032XXX@mn5 is handled outside AS , no?

I think I did not follow everything. Maybe my explanation above helps the last question you had, about being handled outside?

As for what you suggested, where would `~/$your_real_user_platforms.yml` be resolved to? To my actual user home directory, or to the testing suite user home directory? The testing suite cannot access files from users (from a quick test).

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Nov 20, 2024, 13:54

Thanks @kinow 

Okay, then there are two things to change: 

* [ ] the platform.user, that is what I commented

The problem is that autosubmit doesn't read the name from the .ssh/config, so we also need to have the platforms defined with the correct user

> As for what you suggested, where would `~/$your_real_user_platforms.yml` be resolved to? To my actual user home directory, or to the testing suite user home directory? The testing suite cannot access files from users (from a quick test).

Thanks for testing I already had that in mind: 

I know you can't access the home of your original user. 

However, we can just have new files in the folder of the tsuite1 user or any accessible path.

The file name would be different for each original user; we can use: /home/tsuite1/platforms/${SUDO_USER}_platforms.yml. 

Then, Autosubmit would replace the platforms.platform.USER: directive, per the ones you have set there.

We can also simplify the env variable: 

`export ENV_PLATFORMS_PATH="~/platforms/platform_$your_real_user.yml"`

to 

`export ENV_PLATFORMS_PATH="~/platforms/platform_${SUDO_USER}.yml"`

It could even be set automatically in the ~./bashrc of the tsuite1 user. 

* [ ] change what .ssh/config is being read 

```python
paramiko.py: def connect(...)

ssh_config_path = "~/.ssh/config" if not os.getenv("$ENV_PLATFORMS_PATH") else f"~/.ssh/config_{os.getenv("$SUDO_USER")}" # or other condition
self._user_config_file = os.path.expanduser(ssh_config_path)
```

In summary:

# tsuite home

- `/home/tsuite1`
  - `platforms/`
    - `platform_froura.yml`
    - `platform_brunokino.yml`
  - `.ssh/`
    - `config_froura`
    - `config_brunokino`

And the different files: 

```bash
[tsuite1@climatedt-wf ~]$ cat ~/.ssh/config_froura
Host lumi-cluster
 HostName lumi.csc.fi
 User otheruser
Host mn5-cluster1
 HostName glogin1.bsc.es
 User otheruser
Host mn5-cluster2
 HostName glogin2.bsc.es
 User otheruser
Host mn5-cluster3
 HostName glogin3.bsc.es
 User otheruser
```

```bash
[tsuite1@climatedt-wf ~]$ cat ~/.ssh/config_brunokino
Host lumi-cluster
 HostName lumi.csc.fi
 User brunokino
Host mn5-cluster1
 HostName glogin1.bsc.es
 User bsc032841
Host mn5-cluster2
 HostName glogin2.bsc.es
 User bsc032841
Host mn5-cluster3
 HostName glogin3.bsc.es
 User bsc032841
```
- `~/platforms/platform_brunokino.yml`

```yaml
Platforms:
  marenostrum5:
    user: bsc032brunokino
  lumi: 
    ...
```
- `platforms/platform_froura.yml`


```yaml
Platforms:
  marenostrum5:
    user: bsc032froura
  lumi: 
    ...
```

And the env variable: 

`$ export ENV_PLATFORMS_PATH="~/platforms/platform_${SUDO_USER}.yml"`

The minimal.yml files of the experiments of the tsuite1 user must have the POST key set to that env path. ( That would be permanent once it is set once ).

I prefer to have it in the minimal.yml to don't change the default post that is set in the boostrap file

**Comment by dbeltrankyl** [comment]
A test in the climate-dt machine is pending. Aimed for 4.1.13-dev ( just after the 4.1.12 release). 

If I recall correctly, this should be included in the tsuite as soon as 4.1.12 is shipped

**Comment by dbeltrankyl** [comment]
Assigned @kinow for the testing part as I don't have access to that account
